pkg/controllers/nodeclaim/tagging/controller.go

```diff
@@ -0,0 +1,158 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package nodeclaim
+
+import (
+	"context"
+	"regexp"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	controllerruntime "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/aws/karpenter/pkg/apis/v1beta1"
+	"github.com/aws/karpenter/pkg/providers/instance"
+	"github.com/aws/karpenter/pkg/utils"
+
+	corev1beta1 "github.com/aws/karpenter-core/pkg/apis/v1beta1"
+	corecontroller "github.com/aws/karpenter-core/pkg/operator/controller"
+)
+
+type Controller struct {
+	kubeClient       client.Client
+	instanceProvider *instance.Provider
+	ec2              *ec2.EC2
+	defaultTagRgx    *regexp.Regexp
+	timer            *time.Timer
+}
+
+func NewController(kubeClient client.Client, sess *session.Session, instanceProvider *instance.Provider) *Controller {
+	return &Controller{
+		kubeClient:       kubeClient,
+		instanceProvider: instanceProvider,
+		ec2:              ec2.New(sess),
+		defaultTagRgx:    regexp.MustCompile(corev1beta1.NodePoolLabelKey + "/.*"),
+		timer:            time.NewTimer(0),
+	}
+}
+
+func (c *Controller) Name() string {
+	return "ec2-instance-tagger"
+}
+
+func (c *Controller) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	nodeClaim := &corev1beta1.NodeClaim{}
+	if err := c.kubeClient.Get(ctx, req.NamespacedName, nodeClaim); err != nil {
+		return reconcile.Result{}, client.IgnoreNotFound(err)
+	}
+
+	if !isNodeClaimTaggable(nodeClaim) {
+		return reconcile.Result{}, nil
+	}
+
+	// Extract resource ID from node spec
+	// TODO: Should we retry here? If the ProviderID is malformed is there a case where it is later repaired?
+	resourceID, err := utils.ParseInstanceID(nodeClaim.Status.ProviderID)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// Check if instance has already been tagged
+	isTagged, err := c.isInstanceTagged(ctx, resourceID)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	if !isTagged {
+		err = c.tagInstance(nodeClaim, resourceID)
+		if err != nil {
+			return reconcile.Result{Requeue: true}, client.IgnoreNotFound(err)
+		}
+	}
+
+	storedNodeClaim := nodeClaim.DeepCopy()
+	nodeClaim.Annotations[v1beta1.AnnotationInstanceTagged] = "true"
+	err = c.kubeClient.Patch(ctx, nodeClaim, client.MergeFrom(storedNodeClaim))
```

Comment:
We should only patch, if we know that the annotations actually changed here, so we should do something like `if !equality.semantic.DeepEqual` like in the other controllers that perform patch